### PR TITLE
object oriented mode solver

### PR DESCRIPTION
**Pull request by HelgeGehring**

### CODE REVIEW

**Comment by HelgeGehring** [comment]
@simbilod what do you think?

**Comment by simbilod** [comment]
Apparently you need to press "review" for comments to show

File: femwell/mode_solver.py [hunk 1]
```diff
@@ -24,6 +29,105 @@
...
+    def omega(self):
+        return 2 * np.pi * self.frequency
+
+    @property
+    def k0(self):
+        return self.omega / speed_of_light
+
+    @property
+    def wavelength(self):
+        return speed_of_light / self.frequency
```
[the remark this answers is not in the document]
Comment (HelgeGehring) [reply femwell/mode_solver.py:80]: hmm, I'd like a lot to do everything in SI-units, but it seems to lower the resolution. I'll investigate that further.
at the moment, we would use also another unit for frequency if we use non-si-units for wavelength

um -> GHz for example... what would here be the best way? I would want to keep `Mode` the same for electronics and photonics

File: femwell/mode_solver.py [hunk 2]
```diff
@@ -24,6 +29,105 @@
...
+        return ex_sum / (ex_sum + ey_sum)
+
+    @cached_property
+    def tm_fraction(self):
+        return 1 - self.te_fraction
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}(k: {self.k}, n_eff:{self.n_eff})"
+
+    def calculate_overlap(self, mode):
```
[the remark this answers is not in the document]
Comment (HelgeGehring) [reply femwell/mode_solver.py:113]: Yeah, this seems to be a good starting point!
I think the mode solver starts settling, so we can now start with typing and comments.
For type checks we can use `mypy` i think

File: femwell/mode_solver.py [hunk 3]
```diff
@@ -24,6 +29,105 @@
...
+
+@dataclass(frozen=True)
+class Mode:
+    frequency: float
+    k: float
+    basis_epsilon_r: Basis
+    epsilon_r: NDArray
+    basis: Basis
+    E: NDArray
+    H: NDArray
```
[the remark this answers is not in the document]
Comment (HelgeGehring) [reply femwell/mode_solver.py:64]: Yeah, a mode is calculated using an epsilon_r and a basis (which also defines the mesh).
More specifically, compute_modes returns a `Modes` objects, which is more or less a list of `Mode` objects.
There's no reason to change these modes after they are computed (only caching of derived results), so it makes sense, that these modes also provide the basis and the epsilon_r on which they were calculated as it's handy for subsequent calculations

File: femwell/mode_solver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (simbilod) [reply femwell/mode_solver.py:80]: The simplest solution is to go dimensionless like is done in MEEP/MPB, see https://meep.readthedocs.io/en/latest/Introduction/#units-in-meep , or see table here https://github.com/abulnaga1/MPB_Simulations

[the remark this answers is not in the document]
Comment (HelgeGehring) [reply femwell/mode_solver.py:80]: Sounds like a solution :)

My main worry would be, that there would always be a bit of confusion, how to put in numbers and also when it comes to calculation of derived values, e.g. the characteristic impedance.

What do you think? I somehow would like if it keeps close to "reality", but dimensionless is somehow elegant 🤔

Maybe there's also a way to have both?

[the remark this answers is not in the document]
Comment (simbilod) [reply femwell/mode_solver.py:80]: We could keep the mode solver dimensionless, and put all the unit manipulations and conversions outside.

For instance:

import femwell.units as units

# Defaults could be
mode = Mode(frequency=units.frequency(value=193, units="THz", length_units="um"))

where 

units.frequency(value=193, units="THz", length_units="um") = 0.67

but this could also do conversions like

units.frequency(value=193E3, units="GHz", length_units="um") = 0.67

Basically the required conversions happen in that utility module

The same could be done for outputs such as characteristic impedance. 

We could also go through Pint in that module.

[the remark this answers is not in the document]
Comment (HelgeGehring) [reply femwell/mode_solver.py:80]: Sounds great!

The only problem I'd see here: what is the precision of calculations of derived values?

We would probably also need to keep those calculations dimensionless, too?

I think I'll first have a look if we can somehow get the mode solver not caring about dimensions, if that doesn't work, i think this would be a great solution?

[the remark this answers is not in the document]
Comment (HelgeGehring) [reply femwell/mode_solver.py]: Okay, it seems now that we have it scale-invariant - i made sure, that the order of magnitude of E_x/y relative to E_z is scale invariant.
If this works generally, we can just stay with units :)

**Comment by HelgeGehring** [comment]
- [ ] the overlap integrals with scipy as mode solver are wrong currently.

**Comment by HelgeGehring** [comment]
@simbilod at the moment the mode solver is in
`femwell.maxwell.waveguide`
to keep the naming consistant for photonic and electric - would there be a better name than waveguide?

**Comment by simbilod** [comment]
> @simbilod at the moment the mode solver is in `femwell.maxwell.waveguide` to keep the naming consistant for photonic and electric - would there be a better name than waveguide?

In RF, for integrated circuits at least), it's usually known as transmission line. But waveguide is also used (especially for non-integrated 3D structures: https://en.wikipedia.org/wiki/Waveguide#Radio-frequency_waveguides )

**Comment by HelgeGehring** [comment]
Okay, then it seems we probably best stay with the current name as it seems to fit to both? Transmission line sounds somehow wrong for photonics

Another naming could be
`femwell.maxwell.eigenmode` 
or just
`femwell.eigenmode`
?
But it also somehow needs to consistent in the end with the periodic eigenmode solver :thinking: And eigenmode would be quite general :thinking:

**Comment by simbilod** [comment]
Waveguide is great!

**Comment by HelgeGehring** [comment]
Thanks! Then we'll stay with the naming :)
